### PR TITLE
feat(async-flow): error on guest E use

### PR DESCRIPTION
**Pull request by erights**
closes: #XXXX
refs: #9299 #9322

## Description

Prepare for #9322 by making any guest use of `E` until then cause an error. We expect that it might be a while before #9322 is ready for review. By merging this PR soon, we prevent any guest code or logs that would commit us to an incompat way of handling `E`.

### Security Considerations

none
### Scaling Considerations

none
### Documentation Considerations

Should document that guests cannot invoke host vows or remotables with `E` until #9322 , which won't happen immediately.
### Testing Considerations

- [x] need to test what kind of error state this goes into. Should be a panic, so that an upgrade can unblock guest execution that got stuck trying to `E`.
### Upgrade Considerations

The point. By making such use of `E` an error now, we ensure that #9322 can proceed without causing any compat problem with committed durable state.

### CODE REVIEW

File: packages/async-flow/src/replay-membrane.js [hunk 1]
```diff
@@ -258,11 +311,21 @@ export const makeReplayMembrane = (
...
-    guestPromiseMap.set(promise, harden({ resolve, reject }));
+  /**
+   * @param {Vow} hVow
+   * @param {Promise} [promiseKey]
+   *   If provided, use this promise as the key in the guestPromiseMap
+   *   rather than the returned promise. This only happens when the
+   *   promiseKey ends up forwarded to the returned promise anyway, so
+   *   associating it with this resolve/reject pair is not incorrect.
+   *   It is needed when `promiseKey` is also entered into the bijection
+   *   paired with hVow.
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js]: Let's leave the `promiseKey` stuff out of this PR for now.

**Review by mhofman — COMMENTED** [review]
I'll let @michaelfig review the `HandledPromise` logic in the `makeGuest*` functions, my brain always hurt thinking of the handler plumbing.

> * need to test what kind of error state this goes into. Should be a panic, so that an upgrade can unblock guest execution that got stuck trying to `E`.

I just had a pesky thought about the following:

```js
const payment= E(purse).withdraw(amount);
foo.receive(payment); // note not using `E` here. `foo` is a guest wrapper of a host local exo
```

at the time `receive` is called, `payment` is a basic `HandledPromise` unknown to the membrane because the promise handler is always called in a future turn (but always same crank). I honestly do now know how to solve this except providing our own `E` with our own `HandledPromise` implementation, or somehow enforce that the `payment` promise argument becomes know by the next turn, but I'm really warry of going down turn counting routes.

File: packages/async-flow/src/replay-membrane.js (hunk 1 above)
Comment (erights) [reply packages/async-flow/src/replay-membrane.js]: Done

**Comment by erights** [comment]
> I'll let @michaelfig review the `HandledPromise` logic in the `makeGuest*` functions, my brain always hurt thinking of the handler plumbing.
> 
> > * need to test what kind of error state this goes into. Should be a panic, so that an upgrade can unblock guest execution that got stuck trying to `E`.
> 
> I just had a pesky thought about the following:
> 
> ```js
> const payment= E(purse).withdraw(amount);
> foo.receive(payment); // note not using `E` here. `foo` is a guest wrapper of a host local exo
> ```
> 
> at the time `receive` is called, `payment` is a basic `HandledPromise` unknown to the membrane because the promise handler is always called in a future turn (but always same crank). I honestly do now know how to solve this except providing our own `E` with our own `HandledPromise` implementation, or somehow enforce that the `payment` promise argument becomes know by the next turn, but I'm really warry of going down turn counting routes.

I don't understand the problem. You didn't say what `purse` is, so let's break it down into two cases
- `purse` is a guest thing, so the `E(purse).withdraw(amount)` is completely within the guest side. In this case, `payment` is a guest-created guest promise. Then, on the next line when `payment` is used as an argument in a host call, it fails for reasons already tested before this PR: The guest cannot pass guest-created guest promises as arguments in host calls.
- `purse`, like `foo`, is guest wrapper of a host remotable. In that case it fails on this line because of exactly the use of `E` prohibited by this PR.

My guest is you mean the first. But either way, AFAICT, it should work. What am I missing?

**Comment by erights** [comment]
I asked

> What am I missing?

offline @mhofman explained

> basically the problem is that even if `purse` is a guest wrapper, by the time we see `payment` used as an argument, the `purse` "pending / presence handler" has not yet been invoked (delayed by a turn always by eventual send / HandledPromise), so we don't know that `payment` is a handled promise derived from the membrane.

to which I responded

> Right. We’d need to register it during the eventual send itself, which requires some kind of handler to run synchronously, which conflicts with our protection against reentrancy.

> Just to separate issues, this threatens https://github.com/Agoric/agoric-sdk/pull/9322 , but not https://github.com/Agoric/agoric-sdk/pull/9443 , right? The latter (#9443) can still proceed?

File: packages/async-flow/src/replay-membrane.js [hunk 2]
```diff
@@ -199,7 +202,7 @@ export const makeReplayMembrane = (
       default: {
         // @ts-expect-error TS correctly knows this case would be outside
         // the type. But that's what we want to check.
-        throw Fail`unexpected outcome kind ${q(outcome.kind)}`;
+        throw PanicTag`unexpected outcome kind ${q(outcome.kind)}`;
```
Comment (erights) [inline packages/async-flow/src/replay-membrane.js]: This is a drive-by fix

**Comment by erights** [comment]
https://github.com/Agoric/agoric-sdk/blob/ceec1d5655ab909bfa85e312d53cddeee64623f3/packages/async-flow/test/bad-host.test.js#L64-L67

verifies that passing a guest-created promise as argument of a host call is a panic. With that settled, this PR is Ready for re-review

File: packages/async-flow/src/replay-membrane.js [hunk 3]
```diff
@@ -31,6 +34,9 @@ export const makeReplayMembrane = (
 
   let stopped = false;
 
+  const PanicTag = (template, ...args) =>
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js]: Why `PanicTag` and not a simple `Panic` ?

File: packages/async-flow/src/replay-membrane.js [hunk 4]
```diff
@@ -196,9 +202,61 @@ export const makeReplayMembrane = (
...
+    }); // no unfulfilledHandler
+    if (typeof guestPresence !== 'object') {
+      throw Fail`presence expected to be object ${guestPresence}`;
     }
+    assign(guestPresence, fromEntries(methodEntries));
+    const result = Remotable(iface, undefined, guestPresence);
+    result === guestPresence ||
+      Fail`Remotable expected to make presence in place: ${guestPresence} vs ${result}`;
+    return result;
+  };
```
Comment (mhofman) [inline packages/async-flow/src/replay-membrane.js:245]: From my understanding of HandledPromise this seems correct, but I really would be more confident if @michaelfig could double check, as this API is fairly convoluted.

File: packages/async-flow/test/replay-membrane-eventual.test.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  /** @type {Pingee} */
+  const guestPingee = mem.hostToGuest(pingee);
+  t.deepEqual(log.dump(), []);
+
+  await t.throwsAsync(() => E(guestPingee).ping(), {
+    message:
+      'panic over "[Error: guest eventual send not yet supported: \\"[Alleged: Pingee guest wrapper]\\".ping([]) -> \\"[Promise]\\"]"',
+  });
+
+  t.deepEqual(log.dump(), []);
```
Comment (mhofman) [inline packages/async-flow/test/replay-membrane-eventual.test.js]: I'd like to capture in a test that the handler is not synchronously invoked, and this can have effects on the persisted log before the panic occurs.

```suggestion
  const pingTestSendResult = t.throwsAsync(() => E(guestPingee).ping("send"), {
    message:
      'panic over "[Error: guest eventual send not yet supported: \\"[Alleged: Pingee guest wrapper]\\".ping(["send"]) -> \\"[Promise]\\"]"',
  });
  
  guestPingee.ping("call");
  
  await pingTestSendResult;

  t.deepEqual(log.dump(), [
    ['checkCall', pingee, 'ping', ["call"], 1],
    ['doReturn', 1, undefined],
  ]);
```

**Review by mhofman — APPROVED** [review]
Conditional approval assuming the suggested test change is not controversial. I would also still feel more comfortable if @michaelfig could double check that we're setting up eventual-send handlers the right way on guest objects.

File: packages/async-flow/src/replay-membrane.js (hunk 2 above)
Comment (erights) [reply packages/async-flow/src/replay-membrane.js]: Changed to `Panic` as suggested below. Done.

File: packages/async-flow/src/replay-membrane.js (hunk 3 above)
Comment (erights) [reply packages/async-flow/src/replay-membrane.js]: Done

File: packages/async-flow/test/replay-membrane-eventual.test.js (hunk 5 above)
Comment (erights) [reply packages/async-flow/test/replay-membrane-eventual.test.js]: Used your code -- thanks!

Done.

File: packages/async-flow/src/replay-membrane.js (hunk 4 above)
Comment (michaelfig) [reply packages/async-flow/src/replay-membrane.js:245]: Confirmed.  These are the correct contortions given the current HandledPromise API.

**Review by michaelfig — APPROVED** [review]
